src/autoscaler/configutil/cf.go

```diff
@@ -0,0 +1,194 @@
+package configutil
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+
+	"code.cloudfoundry.org/app-autoscaler/src/autoscaler/models"
+	"github.com/cloudfoundry-community/go-cfenv"
+)
+
+var (
+	ErrReadEnvironment   = errors.New("failed to read environment variables")
+	ErrDbServiceNotFound = errors.New("failed to get service by name")
+	ErrMissingCredential = errors.New("failed to get required credential from service")
+)
+
+type VCAPConfigurationReader interface {
+	MaterializeDBFromService(dbName string) (string, error)
+	MaterializeTLSConfigFromService(serviceName string) (models.TLSCerts, error)
+	GetServiceCredentialContent(serviceName string, credentialKey string) ([]byte, error)
+	GetPort() int
+	IsRunningOnCF() bool
+}
+
+type VCAPConfiguration struct {
+	appEnv *cfenv.App
+}
+
+func NewVCAPConfigurationReader() (*VCAPConfiguration, error) {
+	appEnv, err := cfenv.Current()
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", ErrReadEnvironment, err)
+	}
+	return &VCAPConfiguration{appEnv: appEnv}, nil
+}
+
+func (vc *VCAPConfiguration) GetPort() int {
+	return vc.appEnv.Port
+}
+func (vc *VCAPConfiguration) IsRunningOnCF() bool {
+	return cfenv.IsRunningOnCF()
+}
+
+func (vc *VCAPConfiguration) GetServiceCredentialContent(serviceName, credentialKey string) ([]byte, error) {
+	service, err := vc.getServiceByName(serviceName)
+	if err != nil {
+		return []byte(""), err
+	}
+
+	content, ok := service.CredentialString(credentialKey)
+	if !ok {
+		return []byte(""), fmt.Errorf("%w: %s", ErrMissingCredential, credentialKey)
+	}
+
+	return []byte(content), nil
+}
+
+func (vc *VCAPConfiguration) MaterializeTLSConfigFromService(serviceName string) (models.TLSCerts, error) {
+	service, err := vc.getServiceByName(serviceName)
+	if err != nil {
+		return models.TLSCerts{}, err
+	}
+
+	tlsCerts, err := vc.buildTLSCerts(service, serviceName)
+	if err != nil {
+		return models.TLSCerts{}, err
+	}
+
+	return tlsCerts, nil
+}
+
+func (vc *VCAPConfiguration) MaterializeDBFromService(dbName string) (string, error) {
+	service, err := vc.getServiceByName(dbName)
+	if err != nil {
+		return "", err
+	}
+
+	dbURL, err := vc.buildDatabaseURL(service, dbName)
+	if err != nil {
+		return "", err
+	}
+
+	return dbURL.String(), nil
+}
+
+func (vc *VCAPConfiguration) getServiceByName(serviceName string) (*cfenv.Service, error) {
```

Comment:
`getServiceByTag`? Also the various `serviceName` params should probably be `serviceTag`.

Reply:
fixed in 47f5970a9